edb/server/protocol/protocol.pyx

```diff
@@ -189,13 +231,89 @@ cdef class HttpProtocol:
             response.body,
             response.close_connection)
 
+    def _switch_to_binary_protocol(self, data=None):
+        binproto = binary.EdgeConnection(self.server, self.external_auth)
+        self.transport.set_protocol(binproto)
+        binproto.connection_made(self.transport)
+        if data:
+            binproto.data_received(data)
+        return binproto
+
+    def _init_http_parser(self):
+        self.parser = httptools.HttpRequestParser(self)
+        self.current_request = HttpRequest()
+
+    async def _forward_first_data(self, data):
+        # As we stole the "first data", we need to manually send it back to
+        # the SSLProtocol. The hack here is highly-coupled with uvloop impl.
+        transport = self.transport  # The TCP transport
+        for i in range(3):
+            await asyncio.sleep(0)
+            ssl_proto = self.transport.get_protocol()
+            if ssl_proto is not self:
+                break
+        else:
+            raise RuntimeError("start_tls() hasn't run in 3 loop iterations")
+
+        # Delay for one more iteration to make sure the first data is fed after
+        # SSLProtocol.connection_made() is called.
+        await asyncio.sleep(0)
+
+        data_len = len(data)
+        buf = ssl_proto.get_buffer(data_len)
+        buf[:data_len] = data
+        ssl_proto.buffer_updated(data_len)
+
+    async def _start_tls(self):
+        self.transport = await self.loop.start_tls(
+            self.transport, self, self.sslctx, server_side=True
+        )
+        sslobj = self.transport.get_extra_info('ssl_object')
+        if sslobj.selected_alpn_protocol() == 'edgedb-binary':
+            self._switch_to_binary_protocol()
+        else:
+            # It's either HTTP as the negotiated protocol, or the negotiation
+            # failed and we have no idea what ALPN the client has set. Here we
+            # just start talking in HTTP, and let the client bindings decide if
+            # this is an error based on the ALPN result.
+            self._init_http_parser()
+
+    async def _return_binary_error(self, proto):
+        await proto.write_error(errors.BinaryProtocolError(
+            'TLS Required',
```

Comment:
Can we add a test for this condition?  `edgedb-python` auto-downgrades to cleartext on cert errors, we can utilize that.

Reply:
> `edgedb-python` auto-downgrades to cleartext on cert errors

Oh it's actually the opposite - it auto-downgrades to cleartext if the `SSLError` is not a cert error. That means the server is likely an old one without TLS support, and the client binding is allowed to use cleartext.

I just added 2 tests using `edb.protocol.protocol`, that should cover the case.